graphstorm-processing/graphstorm_processing/data_transformations/spark_utils.py

```diff
@@ -0,0 +1,190 @@
+
+    # TODO: These settings shouldn't be necessary for container execution,
+    # can we create such a Spark context only for testing?
+    if not sm_execution and filesystem_type == "s3":
+        spark_builder.config(
+            "spark.jars.packages",
+            "org.apache.hadoop:hadoop-aws:2.10.2," "org.apache.hadoop:hadoop-client:2.10.2",
+        ).config("spark.jars.excludes", "com.google.guava:guava").config(
+            "spark.executor.extraJavaOptions", "-Dcom.amazonaws.services.s3.enableV4=true"
+        ).config(
+            "spark.driver.extraJavaOptions", "-Dcom.amazonaws.services.s3.enableV4=true"
+        )
+
+    spark = spark_builder.getOrCreate()
+
+    spark.sparkContext.setLogLevel("ERROR")
+    logger = spark.sparkContext._jvm.org.apache.log4j
+    logger.LogManager.getLogger("org").setLevel(logger.Level.ERROR)
+    logger.LogManager.getLogger("py4j").setLevel(logger.Level.ERROR)
+    spark_logger = logging.getLogger("py4j.java_gateway")
+    spark_logger.setLevel(logging.ERROR)
+
+    hadoop_config = spark.sparkContext._jsc.hadoopConfiguration()
+    # This is needed to save RDDs which is the only way to write nested Dataframes into CSV format
+    hadoop_config.set(
+        "mapred.output.committer.class", "org.apache.hadoop.mapred.FileOutputCommitter"
+    )
+    # See https://aws.amazon.com/premiumsupport/knowledge-center/emr-timeout-connection-wait/
+    hadoop_config.set("fs.s3.maxConnections", "5000")
+    hadoop_config.set("fs.s3a.connection.maximum", "150")
+    # Only used for local testing and container execution
+    if not sm_execution and filesystem_type == "s3":
+        logging.info("Setting up local Spark instance for S3 access...")
+        hadoop_config.set(
+            "fs.s3a.aws.credentials.provider",
+            "com.amazonaws.auth.DefaultAWSCredentialsProviderChain",
+        )
+        hadoop_config.set("fs.s3.impl", "org.apache.hadoop.fs.s3a.S3AFileSystem")
+        hadoop_config.set("fs.AbstractFileSystem.s3a.imp", "org.apache.hadoop.fs.s3a.S3A")
+        spark.sparkContext.setSystemProperty("com.amazonaws.services.s3.enableV4", "true")
+
+    return spark
+
+
+def safe_rename_column(
+    dataframe: DataFrame, old_colum_name: str, new_column_name: str
+) -> Tuple[DataFrame, str]:
+    """Safely rename a column in a dataframe.
+
+    If the requested column to be renamed does not exist will log a warning.
+
+    If the new name would be the same as another existing column, we modify the requested
+    name and return the new modified value, along with the modified DataFrame.
+
+    Parameters
+    ----------
+    dataframe : DataFrame
+        The dataframe to be modified.
+    old_colum_name : str
+        The name of the column to be renamed.
+    new_column_name : str
+        The new name for the column.
+
+    Returns
+    -------
+    DataFrame
+        The modified dataframe.
+    """
+    if old_colum_name in dataframe.columns:
+        if new_column_name in dataframe.columns:
```

Comment:
We may need to make it a recursive call to avoid t eh case when ```new_column_name + "_2"``` also exists.

Reply:
Instead of adding a recursive call we can add a hash code instead of `_2`. Recursive calls can often lead to issues so I'd rather we avoid adding on here.

Reply:
Then we can use a hash code. either way is fine to me.